crates/vm/src/expr.rs

```diff
@@ -712,80 +738,313 @@ impl Query {
 
 // IndexArgument represents an equality or range predicate that can be answered
 // using an index.
-pub enum IndexArgument {
+#[derive(Debug, PartialEq, Clone)]
+enum IndexArgument<'a> {
     Eq {
-        col_id: ColId,
+        columns: &'a ColList,
         value: AlgebraicValue,
     },
     LowerBound {
-        col_id: ColId,
+        columns: &'a ColList,
         value: AlgebraicValue,
         inclusive: bool,
     },
     UpperBound {
-        col_id: ColId,
+        columns: &'a ColList,
         value: AlgebraicValue,
         inclusive: bool,
     },
 }
 
-// Sargable stands for Search ARGument ABLE.
-// A sargable predicate is one that can be answered using an index.
-fn is_sargable(table: &SourceExpr, op: &ColumnOp) -> Option<IndexArgument> {
-    if let ColumnOp::Cmp {
-        op: OpQuery::Cmp(op),
-        lhs,
-        rhs,
-    } = op
+#[derive(Debug, PartialEq, Clone)]
+enum IndexColumnOp<'a> {
+    Index(IndexArgument<'a>),
+    Scan(&'a ColumnOp),
+}
+
+/// Extracts `name = val` when `lhs` is a field that exists and `rhs` is a value.
+fn ext_field_val<'a>(
+    header: &'a Header,
+    lhs: &'a ColumnOp,
+    rhs: &'a ColumnOp,
+) -> Option<(&'a Column, &'a AlgebraicValue)> {
+    if let (ColumnOp::Field(FieldExpr::Name(name)), ColumnOp::Field(FieldExpr::Value(val))) = (lhs, rhs) {
+        return header.column(name).map(|c| (c, val));
+    }
+    None
+}
+
+/// Extracts `name = val` when `op` is `name = val` and `name` exists.
+fn ext_cmp_field_val<'a>(header: &'a Header, op: &'a ColumnOp) -> Option<(&'a OpCmp, &'a Column, &'a AlgebraicValue)> {
+    match op {
+        ColumnOp::Cmp {
+            op: OpQuery::Cmp(op),
+            lhs,
+            rhs,
+        } => ext_field_val(header, lhs, rhs).map(|(f, v)| (op, f, v)),
+        _ => None,
+    }
+}
+
+fn make_index_arg(cmp: OpCmp, columns: &ColList, value: AlgebraicValue) -> IndexColumnOp<'_> {
+    let arg = match cmp {
+        OpCmp::Eq => IndexArgument::Eq { columns, value },
+        OpCmp::NotEq => unreachable!("using a non-indexed scan for NotEq instead"),
```

Comment:
I think is more clear if says "NotEq have not IndexArgument, filtered-out by caller